### PR TITLE
Support breakpoint evaluation to use VariableTable

### PR DESCRIPTION
**Pull request by hxiong388**
* Created `Snappoint` and `Logpoint` classes. Both inherit from the original `Breakpoint`.
* Refactored `Evaluator` to move a lot of evaluation logic into `Snappoint` and `Logpoint` classes for upcoming evaluation related features
* Introduces `Snappoint` evaluation using `VariableTable` to store compound variables. So repeatedly referenced compound variables can be defined once in `VariableTable` and reused during evaluation.

### CODE REVIEW

File: google-cloud-debugger/lib/google/cloud/debugger/breakpoint/variable_table.rb [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+          ##
+          # @private Array to store variables.
+          attr_accessor :variables
+
+          ##
+          # @private Item in the variables list. :orig_var is reference to the
+          # original Ruby variable. :var is the equivalent
+          # Breakpoint::Variable.
+          ListItem = Struct.new(:orig_var, :var)
```
Comment (dazuma) [inline google-cloud-debugger/lib/google/cloud/debugger/breakpoint/variable_table.rb]: nit: Lose the parentheses

File: google-cloud-debugger/lib/google/cloud/debugger/logpoint.rb [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        # @return [String] The formatted message string
+        #
+        def format_message message_format, expressions
+          # Substitute placeholders with expressions
+          message = message_format.gsub(/(?<!\$)\$\d+/) do |placeholder|
+            index = placeholder.match(/\$(\d+)/)[1].to_i
+            index < expressions.size ? expressions[index].inspect : ""
+          end
+
+          # Unescape "$" charactors
```
Comment (dazuma) [inline google-cloud-debugger/lib/google/cloud/debugger/logpoint.rb]: s/charactors/characters/

File: google-cloud-debugger/lib/google/cloud/debugger/snappoint.rb [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        #   A list of StackFrame objects that represent state of the
+        #   call stack
+        #
+        def eval_call_stack call_stack_bindings
+          call_stack_bindings.each_with_index do |frame_binding, i|
+            frame_info = StackFrame.new.tap do |sf|
+              sf.function = frame_binding.eval("__method__").to_s
+              sf.location = SourceLocation.new.tap do |l|
+                l.path =
+                  frame_binding.eval("::File.absolute_path(__FILE__)")
```
Comment (dazuma) [inline google-cloud-debugger/lib/google/cloud/debugger/snappoint.rb]: nit: parens (this line and next)

File: google-cloud-debugger/lib/google/cloud/debugger/snappoint.rb [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+                l.line = frame_binding.eval("__LINE__")
+              end
+            end
+
+            if i < STACK_EVAL_DEPTH
+              frame_info.locals =
+                frame_binding.local_variables.map do |local_var_name|
+                  local_var = frame_binding.local_variable_get(local_var_name)
```
Comment (dazuma) [inline google-cloud-debugger/lib/google/cloud/debugger/snappoint.rb]: nit: parens

File: google-cloud-debugger/lib/google/cloud/debugger/snappoint.rb (hunk 3 above)
Comment (hxiong388) [inline google-cloud-debugger/lib/google/cloud/debugger/snappoint.rb]: Fixed

File: google-cloud-debugger/lib/google/cloud/debugger/snappoint.rb (hunk 4 above)
Comment (hxiong388) [inline google-cloud-debugger/lib/google/cloud/debugger/snappoint.rb]: Fixed

File: google-cloud-debugger/lib/google/cloud/debugger/logpoint.rb (hunk 2 above)
Comment (hxiong388) [inline google-cloud-debugger/lib/google/cloud/debugger/logpoint.rb]: Fixed

File: google-cloud-debugger/lib/google/cloud/debugger/breakpoint/variable_table.rb (hunk 1 above)
Comment (hxiong388) [inline google-cloud-debugger/lib/google/cloud/debugger/breakpoint/variable_table.rb]: Fixed

**Pull request merged**